pk/ecc/ecc_sign.c

```diff
@@ -0,0 +1,350 @@
+    if (ret != MP_OKAY) {
+        wc_ecc_free(&ecc);
+        wc_FreeRng(&rng);
+        return ret;
+    }
+    ret = mp_init(&s);
+    if (ret != MP_OKAY) {
+        mp_clear(&r);
+        wc_ecc_free(&ecc);
+        wc_FreeRng(&rng);
+        return ret;
+    }
+
+    /* Import private key "k" */
+    ret = wc_ecc_import_private_key_ex(
+        key, keySz, /* private key "d" */
+        NULL, 0,    /* public (optional) */
+        &ecc,
+        curveId     /* ECC Curve Id */
+    );
+    if (ret == 0) {
+        /* Verify ECC Signature */
+        ret = wc_ecc_sign_hash_ex(
+            hash, hashSz, /* computed hash digest */
+            &rng, &ecc,   /* random and key context */
+            &r, &s        /* r/s as mp_int */
+        );
+
+        /* export r/s */
+        mp_to_unsigned_bin(&r, sig);
+        mp_to_unsigned_bin(&s, sig + curveSz);
+    }
+    
+    mp_clear(&r);
+    mp_clear(&s);
+    wc_ecc_free(&ecc);
+    wc_FreeRng(&rng);
+
+    return ret;
+}
+#endif /* HAVE_ECC_SIGN */
+#endif /* HAVE_ECC */
+
+static void print_hex(uint8_t* data, int sz)
+{
+    int i;
+    for (i = 0; i < sz; i++) {
+        printf("%02X ", data[i]);
+        if (i > 0 && ((i+1) % 16) == 0)
+            printf("\n");
+    }
+    printf("\n");
+}
+
+int main()
+{
+#if defined(HAVE_ECC) && defined(HAVE_ECC_SIGN) && defined(HAVE_ECC_VERIFY) && \
+    !defined(NO_SHA256)
+    int ret;
+    uint8_t hash[WC_SHA256_DIGEST_SIZE];
+    uint8_t sig[ECC_CURVE_SZ*2];
+    uint32_t sigSz = sizeof(sig);
+
+#ifdef DEBUG_WOLFSSL
+    wolfSSL_Debugging_ON();
+#endif
+
+    printf("Running NIST P-256,SHA-256 Sign Test\n");
+
+    ret = crypto_sha256(
+        kMsg, sizeof(kMsg), /* input message */
+        hash, sizeof(hash), /* hash digest result */
+        32                  /* configurable block / chunk size */
+    );
+    if (ret == 0) {
+        /* Sign hash using private key */
+        /* Note: result of an ECC sign varies for each call even with same 
+            private key and hash. This is because a new random public key is 
+            used for each operation. */ 
+        ret = crypto_ecc_sign(
+            kPrivKey, sizeof(kPrivKey), /* private key */
+            hash, sizeof(hash),         /* computed hash digest */
+            sig, &sigSz,                /* signature r/s */
+            ECC_CURVE_SZ,               /* SECP256R1 curve size in bytes */
+            ECC_CURVE_ID                /* curve id */
+        );
+    }
+
+    if (ret == 0) {
+        /* Verify generated signature is valid */
+        ret = crypto_ecc_verify(
+            kPubKey, sizeof(kPubKey),   /* public key point x/y */
+            hash, sizeof(hash),         /* computed hash digest */
+            sig, sigSz,                 /* signature r/s */
+            ECC_CURVE_SZ,               /* curve size in bytes */
+            ECC_CURVE_ID                /* curve id */
+        );
+    }
+
+    printf("Signature %d\n", sigSz);
+    print_hex(sig, sigSz);
```

Comment:
What happens if crypt_sha256 on line 306 is called but returns non-zero and the if conditions on 311 and 325 evaluate to false and we call print_hex with sig uninitialized?